FWCore/Framework/interface/stream/ThinningProducer.h

```diff
@@ -19,13 +19,52 @@
 #include "FWCore/Utilities/interface/propagate_const.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "DataFormats/Common/interface/DetSetVectorNew.h"
 
 #include <memory>
 
 namespace edm {
 
   class EventSetup;
 
+  //default implementation which works with std::vector type collections or those with compatible interfaces
+  template <typename Collection, typename Selector>
+  void fillCollectionForThinning(edm::Handle<Collection> const& input,
+                                 Collection& output,
+                                 ThinnedAssociation& assoc,
+                                 edm::propagate_const<std::unique_ptr<Selector>>& selector) {
+    unsigned int iIndex = 0;
+    for (auto iter = input->begin(), iterEnd = input->end(); iter != iterEnd; ++iter, ++iIndex) {
+      if (selector->choose(iIndex, *iter)) {
+        output.push_back(*iter);
+        assoc.push_back(iIndex);
+        selector->modify(output.back());
+      }
+    }
+  }
+
+  //specialization for edmNew::DetSetVector
+  template <typename T, typename Selector>
+  void fillCollectionForThinning(edm::Handle<edmNew::DetSetVector<T>> const& input,
+                                 edmNew::DetSetVector<T>& output,
```

Comment:
I'd like to avoid (non-test) code in `FWCore` to depend on `DetSetVector`, but there might be a way to avoid it.

Reply:
Ok, open to suggestions here.

Reply:
Is your objection the including of the DetSetVectorNew.h header, or rather the dependency on the interface?

Reply:
Both. The infrastructure should be extensible so that `DetSetVector`-dependent code would be placed in the same place (header) as the `DetSetVector` (or anything else special) is defined. I have an idea how it could be achieved.